symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
+    tasks = []
+    connector = TCPConnector(limit=100, limit_per_host=20)
+    async with ClientSession(
+        loop=loop, timeout=ClientTimeout(total=TIMEOUT), connector=connector
+    ) as client:
+        for filename, debug_id, code_file, code_id, has_code in modules:
+            tasks.append(fetch_and_write(output, client, filename, debug_id))
+            if has_code:
+                tasks.append(fetch_and_write(output, client, code_file, code_id))
+
+        return await asyncio.gather(*tasks)
+
+
+def fetch_all(output, modules):
+    res = asyncio.run(fetch_all_helper(output, modules))
+    res = iter(res)
+    fetched_modules = []
+    for filename, debug_id, code_file, code_id, has_code in modules:
+        fetched_pdb = next(res)
+        if has_code:
+            has_code = next(res)
+        if fetched_pdb:
+            fetched_modules.append((filename, debug_id, code_file, code_id, has_code))
+    return fetched_modules
+
+
+def get_base_data(url):
+    async def helper(url):
+        return await asyncio.gather(
+            fetch_missing_symbols(url),
+            # Symbols that we know belong to us, so don't ask Microsoft for them.
+            get_list("blacklist.txt", "Blacklist"),
+            # Symbols that we know belong to Microsoft, so don't skiplist them.
+            get_list("known-microsoft-symbols.txt", "Known Microsoft symbols"),
+            # Symbols that we've asked for in the past unsuccessfully
+            get_skiplist(),
+        )
+
+    return asyncio.run(helper(url))
+
+
+def gen_zip(output, output_dir, file_index):
+    if not file_index:
+        return
+
+    with zipfile.ZipFile(output, "w", zipfile.ZIP_DEFLATED) as z:
         for f in file_index:
-            z.write(os.path.join(symbol_path, f), f)
-    log.info('Wrote zip as %s' % args.zip)
+            z.write(os.path.join(output_dir, f), f)
+    log.info(f"Wrote zip as {output}")
+
+
+def retry_run(cmd, *arg):
+    for _ in range(5):
+        try:
+            return cmd(*arg)
+        except OSError as e:
+            # Sometimes for any reasons we've "Too much open files"
+            # So wait and try again
+            log.exception(e)
+            time.sleep(10)
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Fetch missing symbols from Microsoft symbol server"
+    )
+    parser.add_argument(
+        "--missing-symbols",
+        type=str,
+        help="missing symbols URL",
+        default=MISSING_SYMBOLS_URL,
+    )
+    parser.add_argument("zip", type=str, help="output zip file")
+    parser.add_argument(
+        "--dump-syms",
+        type=str,
+        help="dump_syms path",
+        default=os.environ.get("DUMP_SYMS_PATH"),
+    )
+
+    args = parser.parse_args()
+
+    assert bool(args.dump_syms), "dump_syms path is empty"
+
+    logging.basicConfig(level=logging.DEBUG)
+    aiohttp_logger = logging.getLogger("aiohttp.client")
+    aiohttp_logger.setLevel(logging.DEBUG)
```

Comment:
Normally we should probably set this to INFO, and just manually set it to DEBUG for development